bfvmm/include/vmcs/vmcs_intel_x64.h

```diff
@@ -3453,187 +3380,794 @@ namespace guest_es_access_rights
+        is_enabled()
+        { return (vm::read(addr, name) & mask) != 0; }
 
-    inline auto get()
-    { return vm::read(addr, name); }
+        template <class T> void set(T val)
+        { set_vm_control(val, msr_addr, addr, name, mask); }
 
-    template<class T> void set(T val)
-    { vm::write(addr, val, name); }
+        template <class T> void set_if_allowed(T val, bool verbose) noexcept
+        { set_vm_control_if_allowed(val, msr_addr, addr, name, mask, verbose); }
 
-    inline bool exists() noexcept
-    { return true; }
+        inline void enable()
+        { set(1UL); }
 
-    namespace type
-    {
-        constexpr const auto mask = 0x000000000000000FUL;
-        constexpr const auto from = 0;
-        constexpr const auto name = "type";
+        inline void disable()
+        { set(0UL); }
 
-        inline auto get()
-        { return (vm::read(addr, name) & mask) >> from; }
+        inline void enable_if_allowed(bool verbose = false) noexcept
+        { set_if_allowed(1UL, verbose); }
 
-        template<class T> void set(T val)
-        { vm::write(addr, (vm::read(addr, name) & ~mask) | ((val << from) & mask), name); }
+        inline void disable_if_allowed(bool verbose = false) noexcept
+        { set_if_allowed(0UL, verbose); }
     }
 
-    namespace s
+    namespace virtualize_x2apic_mode
     {
         constexpr const auto mask = 0x0000000000000010UL;
         constexpr const auto from = 4;
-        constexpr const auto name = "s";
+        constexpr const auto name = "virtualize_x2apic_mode";
 
-        inline auto get()
-        { return (vm::read(addr, name) & mask) >> from; }
+        inline bool
+        is_enabled()
+        { return (vm::read(addr, name) & mask) != 0; }
 
-        template<class T> void set(T val)
-        { vm::write(addr, (vm::read(addr, name) & ~mask) | ((val << from) & mask), name); }
-    }
+        template <class T> void set(T val)
+        { set_vm_control(val, msr_addr, addr, name, mask); }
 
-    namespace dpl
-    {
-        constexpr const auto mask = 0x0000000000000060UL;
-        constexpr const auto from = 5;
-        constexpr const auto name = "dpl";
+        template <class T> void set_if_allowed(T val, bool verbose) noexcept
+        { set_vm_control_if_allowed(val, msr_addr, addr, name, mask, verbose); }
 
-        inline auto get()
-        { return (vm::read(addr, name) & mask) >> from; }
+        inline void enable()
+        { set(1UL); }
+
+        inline void disable()
+        { set(0UL); }
+
+        inline void enable_if_allowed(bool verbose = false) noexcept
+        { set_if_allowed(1UL, verbose); }
+
+        inline void disable_if_allowed(bool verbose = false) noexcept
+        { set_if_allowed(0UL, verbose); }
+    }
+
+    namespace enable_vpid
+    {
+        constexpr const auto mask = 0x0000000000000020UL;
+        constexpr const auto from = 5;
+        constexpr const auto name = "enable_vpid";
+
+        inline bool
+        is_enabled()
```

Comment:
Do you think the `get` should throw an exception?  Something like
```cpp
namespace pin_based_vm_execution_controls
{
    inline auto get()
    { 
        if (!exists())
            throw std::logic_error("pin based controls don't exist");
     
        return vm::read(addr, name);
    }

    inline auto get_if_allowed(bool verbose = false)
    {
        if (!exists() && verbose)
            bfwarning << name << " doesn't exist" << '\n';
 
        if (exists())
            return vm::read(addr, name);
    }
}
```
Another question...do you think we need these for every field, or just the ones that may not exist? I'm pretty sure the pin based controls and primary controls are going to exist on every processor we support, so in those cases I think a  `get_if_allowed` would be redundant

Also we should probably make a `set_if_allowed` to go along with `get_if_allowed` as well.